builtin/providers/azurerm/resource_arm_virtual_network_gateway.go

```diff
@@ -0,0 +1,685 @@
+
+	resGroup, name, err := resourceGroupAndVirtualNetworkGatewayFromId(d.Id())
+	if err != nil {
+		return err
+	}
+
+	resp, err := client.Get(resGroup, name)
+	if err != nil {
+		if resp.StatusCode == http.StatusNotFound {
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error making Read request on VirtualNetwork Gateway %s: %s", name, err)
+	}
+
+	gw := *resp.VirtualNetworkGatewayPropertiesFormat
+
+	d.Set("name", resp.Name)
+	d.Set("resource_group_name", resGroup)
+	d.Set("location", azureRMNormalizeLocation(*resp.Location))
+	d.Set("type", string(gw.GatewayType))
+	d.Set("enable_bgp", gw.EnableBgp)
+	d.Set("active_active", gw.ActiveActive)
+	d.Set("sku", schema.NewSet(hashVirtualNetworkGatewaySku, flattenArmVirtualNetworkGatewaySku(gw.Sku)))
+
+	if string(gw.VpnType) != "" {
+		d.Set("vpn_type", string(gw.VpnType))
+	}
+
+	if gw.GatewayDefaultSite != nil {
+		d.Set("default_local_network_gateway_id", gw.GatewayDefaultSite.ID)
+	}
+
+	d.Set("ip_configuration", flattenArmVirtualNetworkGatewayIPConfigurations(gw.IPConfigurations))
+
+	if gw.VpnClientConfiguration != nil {
+		vpnConfigFlat := flattenArmVirtualNetworkGatewayVpnClientConfig(gw.VpnClientConfiguration)
+		d.Set("vpn_client_configuration", schema.NewSet(hashVirtualNetworkGatewayVpnClientConfig, vpnConfigFlat))
+	}
+
+	if gw.BgpSettings != nil {
+		bgpSettingsFlat := flattenArmVirtualNetworkGatewayBgpSettings(gw.BgpSettings)
+		d.Set("bgp_settings", schema.NewSet(hashVirtualNetworkGatewayBgpSettings, bgpSettingsFlat))
+	}
+
+	flattenAndSetTags(d, resp.Tags)
+
+	return nil
+}
+
+func resourceArmVirtualNetworkGatewayDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient)
+	vnetGatewayClient := client.vnetGatewayClient
+
+	resGroup, name, err := resourceGroupAndVirtualNetworkGatewayFromId(d.Id())
+	if err != nil {
+		return err
+	}
+
+	_, err = vnetGatewayClient.Delete(resGroup, name, make(chan struct{}))
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] Waiting for VirtualNetworkGateway (%s) to be removed", name)
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{"Accepted", "Deleting"},
+		Target:  []string{"NotFound"},
+		Refresh: virtualNetworkGatewayStateRefreshFunc(client, resGroup, name, true),
+		Timeout: 15 * time.Minute,
+	}
+	if _, err := stateConf.WaitForState(); err != nil {
+		return fmt.Errorf("Error waiting for VirtualNetworkGateway (%s) to be removed: %s", name, err)
+	}
+
+	// Gateways are not fully cleaned up when the API indicates the delete operation
+	// has finished, this workaround was suggested by Azure support to avoid conflicts
+	// when modifying/deleting the related subnet or network. Although the API indicated
+	// that the Virtual Network Gateway does not exist anymore, there is still a link
+	// to the Gateway Subnet it has been associated with. This causes an error when
+	// trying to delete the Gateway Subnet immediately after the Virtual Network Gateway
+	// has been deleted.
+
+	d.SetId("")
```

Comment:
Regarding this issue, I tried some of the alternatives which are definitly better approaches before "relying" on retries of subnet deletion attempts.

Regaring the approach to "poll for the deletion of the Virtual Network Gateway using a StateFunc and a ContinuousTargetOccurence value":  I tried exactly this in the gateway resource, but due to the consistency bug the API responds with a 400 error, as if the gateway would have never existed. In this case, the "hidden link" between the gateway subnet and the gateway still exists.

I tried polling using vnetGatewayClient.Get only. I will check the other methods you metioned before: "see if the Virtual Network Gateway is visible through the either Virtual Network Gateway's List method / Resource Group's contents".